text/0931-template-compiler-api.md

```diff
@@ -0,0 +1,431 @@
+const Headline = template("<h1>{{yield}}</h1>");
+
+// local variable access works the same as in current precompileTemplate
+const Section = template(
+  "<Headline>{{@title}}</Headline>", 
+  { 
+    scope: () => ({ Headline}) 
+  }
+);
+
+// in class member position, we can also put private fields in scope
+class extends Component {
+  static {
+    template(
+
+      "<Section @title={{this.title}} @subhead={{this.#secret}} />", 
+      {
+        scope: (instance) => ({ 
+          Section, 
+          "#secret": instance.#secret
+        }),
+      },
+      this
+    )
+  }
+}
+```
+
+> This RFC is focused on making sure the scope accessors can do everything javascript can do, which is why we're including private fields. However, additional work beyond this RFC is required to make the template compiler correctly parse expressions like `{{this.#secret}}`. 
+
+#### Implicit Form
+
+The "Implicit Form" is cheaper and easier to produce because it doesn't need to do any lexical analysis and doesn't need to parse the handlebars at all.
+
+The downside is that data flow is not all statically visible, because it relies on `eval`.
+
+Implicit Form has two key use cases:
+ - as an intermediate format between a preprocessor stage (which can eliminate all GJS special syntax and semantics) and the rest of a standard Javascript toolchain.
+ - as the implementation format in sandbox-like environments where dynamic code execution is the whole point.
+
+Examples of Implicit Form:
+
+```js
+import { template } from '@ember/template-compiler';
+
+// Notice that all of these have the exact same
+// `params` argument. It's always the same. 
+// That's why it's easy to produce.
+
+const Headline = template(
+  "<h1>{{yield}}</h1>",
+  {
+    eval() { return eval(arguments[0]); }
+  }
+);
+
+const Section = template(
+  "<Headline>{{@title}}</Headline>", 
+  { 
+    eval() { return eval(arguments[0]); }
+  }
+);
+
+class extends Component {
+  static {
+    template(
+      "<Section @title={{this.title}} @subhead={{this.#secret}} />", 
+      {
+        // this handles private fields just fine,
+        // see Appendix A.
+        eval() { return eval(arguments[0]); }
+      },
+      this
+    )
+  }
+}
+```
+
+> _eval seems bad, what about Content Security Policy?_<br>
+> Typical apps never needs to actually _run_ the eval. This is a communication format between layers of build tooling. You _can_ run it, if you're making something like an interactive development sandbox. But that is a case that already requires `eval` anyway.
+
+> _Why `arguments[0]` instead of an explicit argument?_<br>
+> If we picked a local name to use for the argument, we would shadow that name in the surrounding scope. Whereas `arguments` is already a keyword that exists for this purpose, and can never collide with other local bindings.
+
+
+### Type Signature
+
+```ts
+import { ComponentLike } from '@glimmer/template';
```

Comment:
Is the intent here to ~upstream the type from `@glint/template`,  or otherwise to provide it from a new `@glimmer/template` package?

Reply:
Sorry, just a typo, that was supposed to say `@glint/template`.

But also: is that OK? I understand that a lot of that package calls itself private, but it also seems like the best type we have for this purpose.

Reply:
Ah, very good. Yes, that specific import is explicitly public and intended for this sort of usage! The only concern we will want to think about is that it forces a consideration for co-evolution of these packages, but that’s largely already true and we also expect the public API here to be quite stable so that’s not a particularly significant concern.